From ebcc4f18d659a4903b58c5f0f35801a8bcb446fe Mon Sep 17 00:00:00 2001
From: Andreas Schneider <asn@cryptomilk.org>
Date: Tue, 11 Aug 2026 08:47:56 +0000
Subject: [PATCH 4/7] On Samsung devices, we need to tell Audio HAL if we're
 running narrow band or wide band

[ported to lineage-23.2/A16: arrow-switch syntax adapted; WBS_NO/NONE arm
converted to block to hold the added statement]

Change-Id: I7802b7a29c017a2cd7018e82772183df1dfa0b89
---
 .../android/bluetooth/hfp/HeadsetStateMachine.java    | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/android/app/src/com/android/bluetooth/hfp/HeadsetStateMachine.java b/android/app/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
index 8978797a3d..baf7ce1687 100644
--- a/android/app/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
+++ b/android/app/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
@@ -166,6 +166,7 @@ class HeadsetStateMachine extends StateMachine {
     private boolean mHasWbsEnabled = false;
     private boolean mHasSwbLc3Enabled = false;
     private boolean mHasSwbAptXEnabled = false;
+    private int mReportScoSampleRate = 0;
     // AT Phone book keeps a group of states used by AT+CPBR commands
     @VisibleForTesting final AtPhonebook mPhonebook;
     // HSP specific
@@ -280,6 +281,7 @@ class HeadsetStateMachine extends StateMachine {
         mHasNrecEnabled = false;
         mHasSwbLc3Enabled = false;
         mHasSwbAptXEnabled = false;
+        mReportScoSampleRate = 0;
     }
 
     public void dump(StringBuilder sb) {
@@ -586,6 +588,7 @@ class HeadsetStateMachine extends StateMachine {
             mHasSwbLc3Enabled = false;
             mHasNrecEnabled = false;
             mHasSwbAptXEnabled = false;
+            mReportScoSampleRate = 0;
             mHasRfcommConnectionCompleted = false;
 
             if (mHeadsetService.mPendingScoConnectionDevice != null
@@ -1828,6 +1831,9 @@ class HeadsetStateMachine extends StateMachine {
             am.setParameters("bt_swb=" + (mHasSwbAptXEnabled ? "0" : "65535"));
         }
         am.setBluetoothHeadsetProperties(getCurrentDeviceName(), mHasNrecEnabled, mHasWbsEnabled);
+        if (mReportScoSampleRate > 0) {
+            am.setParameters("g_sco_samplerate=" + mReportScoSampleRate);
+        }
     }
 
     @VisibleForTesting
@@ -1982,12 +1988,15 @@ class HeadsetStateMachine extends StateMachine {
         switch (wbsConfig) {
             case HeadsetHalConstants.BTHF_WBS_YES -> {
                 mHasWbsEnabled = true;
+                mReportScoSampleRate = 16000;
                 if (mHeadsetService.isAptXSwbEnabled()) {
                     mHasSwbAptXEnabled = false;
                 }
             }
-            case HeadsetHalConstants.BTHF_WBS_NO, HeadsetHalConstants.BTHF_WBS_NONE ->
+            case HeadsetHalConstants.BTHF_WBS_NO, HeadsetHalConstants.BTHF_WBS_NONE -> {
                     mHasWbsEnabled = false;
+                    mReportScoSampleRate = 8000;
+            }
             default -> {
                 Log.e(TAG, "processWBSEvent: unknown wbsConfig " + wbsConfig);
                 return;
